drv/transceivers-server/src/udp.rs

```diff
@@ -0,0 +1,503 @@
+    Reset(ModuleId),
+    Status(ModuleId),
+    Read(ModuleId, MemoryRead),
+    Write(ModuleId, MemoryWrite),
+    ManagementInterface(ManagementInterface),
+    UnexpectedHostResponse(HostResponse),
+    GotSpRequest,
+    GotSpResponse,
+    WrongVersion(u8),
+}
+
+ringbuf!(Trace, 16, Trace::None);
+
+////////////////////////////////////////////////////////////////////////////////
+
+/// Convert from the over-the-network type to our local FPGA type
+fn get_fpga(m: ModuleId) -> Result<FpgaController, Error> {
+    match m.fpga_id {
+        0 => Ok(FpgaController::Left),
+        1 => Ok(FpgaController::Right),
+        i => Err(Error::InvalidFpga(i)),
+    }
+}
+
+/// Convert from the over-the-network type to our local port mask type
+fn get_mask(m: ModuleId) -> Result<FpgaPortMasks, Error> {
+    let fpga_ports: u16 = m.ports.0;
+    match m.fpga_id {
+        0 => Ok(FpgaPortMasks {
+            left: fpga_ports,
+            right: 0,
+        }),
+        1 => Ok(FpgaPortMasks {
+            left: 0,
+            right: fpga_ports,
+        }),
+        i => Err(Error::InvalidFpga(i)),
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+impl ServerImpl {
+    /// Attempt to read and handle data from the `net` socket
+    pub fn check_net(
+        &mut self,
+        rx_data_buf: &mut [u8],
+        tx_data_buf: &mut [u8],
+    ) {
+        const SOCKET: SocketName = SocketName::transceivers;
+
+        match self.net.recv_packet(
+            SOCKET,
+            LargePayloadBehavior::Discard,
+            rx_data_buf,
+        ) {
+            Ok(mut meta) => {
+                // Modify meta.size based on the output packet size
+                let out_len = match hubpack::deserialize(rx_data_buf) {
+                    Ok((msg, data)) => {
+                        self.handle_message(msg, data, tx_data_buf)
+                    }
+                    Err(e) => {
+                        // At this point, deserialization has failed, so we
+                        // can't handle the packet.  We'll attempt to
+                        // deserialize *just the header* (which should never
+                        // change), in the hopes of logging a more detailed
+                        // error message about a version mismatch.
+                        ringbuf_entry!(Trace::DeserializeError(e));
+                        match hubpack::deserialize::<Header>(rx_data_buf) {
+                            Ok((header, _)) => {
+                                ringbuf_entry!(Trace::WrongVersion(
+                                    header.version
+                                ));
+                            }
+                            Err(e) => {
+                                ringbuf_entry!(Trace::DeserializeHeaderError(
+                                    e
+                                ));
+                            }
+                        }
+                        None
```

Comment:
Should the SP send back a protocol error message here? Maybe only in the "`header` deserialized successfully" branch, because then we have the message ID to tell the host it's not going to get an answer.